File: problems_0-1000/125-ValidPalindrome/go/solution.go
package main

func isPalindrome(s string) bool {
	var cleaned []byte
	for _, c := range s {
		if (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') {
			cleaned = append(cleaned, byte(c))
		} else if c >= 'A' && c <= 'Z' {
			cleaned = append(cleaned, byte(c+32))
		}
	}

	left, right := 0, len(cleaned)-1
	for left < right {
		if cleaned[left] != cleaned[right] {
			return false
		}
		left++
		right--
	}

	return true
}

func main() {
	isPalindrome("A man, a plan, a canal: Panama")
}

File: problems_0-1000/125-ValidPalindrome/python/solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        cleaned = []
        for c in s:
            if 'a' <= c <= 'z' or '0' <= c <= '9':
                cleaned.append(c)
            elif 'A' <= c <= 'Z':
                cleaned.append(chr(ord(c) + 32))

        left, right = 0, len(cleaned) - 1
        while left < right:
            if cleaned[left] != cleaned[right]:
                return False
            left += 1
            right -= 1
        return True


if __name__ == '__main__':
    s = Solution()
    s.isPalindrome("A man, a plan, a canal: Panama")
    # print(s.isPalindrome("A man, a plan, a canal: Panama"))